core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java

```diff
@@ -187,7 +186,7 @@ protected NodesAndVersions buildAllocationIdBasedNodes(ShardRouting shard, boole
             }
 
             if (nodeShardState.storeException() == null) {
-                if (allocationId == null && nodeShardState.version() != -1) {
+                if (allocationId == null && nodeShardState.legacyVersion() != -1) {
```

Comment:
One scenario could be: If the shard never made it to started in v3.0.0, and we later restore that shard from a snapshot, it's local data can still be used to speed up restore process.
It's a super-edge case though (and I cannot come up with another one), I'm ok with removing it.